cpp/src/aztec3/circuits/rollup/base/nullifier_tree_testing_harness.cpp

```diff
@@ -10,13 +11,74 @@ NullifierMemoryTreeTestingHarness::NullifierMemoryTreeTestingHarness(size_t dept
     : NullifierMemoryTree(depth)
 {}
 
+// Check for a larger value in an array
+bool check_has_lesser_value(std::vector<fr> const& values, fr const& value)
+{
+    // Must perform comparisons on integers
+    uint256_t value_as_uint = uint256_t(value);
+    for (auto const& v : values) {
+        if (uint256_t(v) < value_as_uint) {
+            return true;
+        }
+    }
+    return false;
+}
+
+// TODO: test
+fr NullifierMemoryTreeTestingHarness::append_value(fr const& value)
+{
+    // If the value is 0, then we force insert the value to increase the size of the tree
```

Comment:
sorry why do we do this again?

Reply:
It allows the size of the tree to progress each kernel circuit so that subtrees can be inserted at once.

Say, for example we are adding three nullifiers (a,b,c), and want to perform a batch insertion of 4 notes each time (in reality its 8 but 4 is quicker to illustrate).

    h(ab,c0)
   /     \
h(a,b) h(c,0)
 / \    /  \
a   b   c   0

To insert this subtree we need to insert 4 values into the tree, increasing its size to 4. If we didnt the next subtree would write to our first 0. 

e.g. if inserting (d,e,f,g)

        h(abcd,efg0)
       /              \
    h(ab,cd)        h(ef,g0)
   /     \         /      \  
h(a,b) h(c,d)    h(e,f) h(g,0)
 / \    /  \    /   \   /   \
a   b   c   d   e   f   g   0

As you can see here we cant cleanly insert our subtree d e f g as we have a leaf that is updating the left subtree. We now must do a few more checks. The hash h(c,0) -> h(c,d) and cascades upwards meaning we have alot more nodes that have changed.

        h(abc0,defg)
       /              \
    h(ab,c0)        h(de,fg)
   /     \         /      \  
h(a,b) h(c,0)    h(d,e) h(f,g)
 / \    /  \    /   \   /   \
a   b   c   0   d   e   f   g

In this case where we allow sparse insertions we can cleanly insert defg without changing any of the leftmost subtree paths.